src/disk_util.hpp

```diff
@@ -0,0 +1,189 @@
+// Copyright 2018 Chia Network Inc
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//    http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SRC_CPP_DISK_UTIL_HPP_
+#define SRC_CPP_DISK_UTIL_HPP_
+
+#include <string>
+#include <sstream>
+#include <fstream>
+#include <iostream>
+
+#ifndef _WIN32
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/file.h>
+#include <unistd.h>
+#include <limits.h>
+#include <stdlib.h>
+#endif
+
+#if !defined(_WIN32) && !defined(__APPLE__)
+#include <sys/sysmacros.h>
+#endif
+
+#include "util.hpp"
+
+namespace DiskUtil {
+
+    // For SSDs parallel writing is preferred.
+    // For HDDs it's slow and causes fragmentation.
+    static inline bool IsParallelWritingPreferred(const std::string &dir)
+    {
+#if defined(__APPLE__) || defined(_WIN32)
+        return true;
+#else
+        struct stat s;
+        memset(&s, 0, sizeof(s));
+
+        if (0 != stat(dir.c_str(), &s)) {
+            std::ostringstream err;
+            err << "Unable to find device name for dir " << dir << ": "
+                << strerror(errno);
+            throw std::runtime_error(err.str());
+        }
+
+        int major_id = major(s.st_dev);
+        int minor_id = 0; // for /dev/sda1, check /dev/sda
+
+        std::ostringstream os;
+        os << "/sys/dev/block/" << major_id << ":" << minor_id
+            << "/queue/rotational";
+        std::string filename = os.str();
+
+        std::ifstream file;
+        file.open(filename.c_str());
+
+        if (file.fail()) {
+            std::ostringstream err;
+            err << "Unable to open " << filename << " for reading: "
+                << strerror(errno);
+            throw std::runtime_error(err.str());
+        }
+
+        std::string line;
+        getline(file, line);
+
+        file.close();
+
+        int is_rotational = 1;
+        if (!line.empty() && line.front() == '0') {
+            is_rotational = 0;
+        }
+
+        return !is_rotational;
+#endif
+    }
+
+    inline bool ShouldLock(const std::string &dir) {
+        try {
+            return !DiskUtil::IsParallelWritingPreferred(dir);
+        } catch (const std::runtime_error &err) {
+            std::cerr << err.what() << std::endl;
+            return false;
+        }
+    }
+}
+
+class DirectoryLock
```

Comment:
you would probably want this class to be non-copyable and non-movable (or movable, but you'd have to make it transfer ownership of the lock).

```
DirectoryLock(DirectoryLock const&) = delete;
```

Reply:
Fixed.